fastcrypto-tbls/src/dkg.rs

```diff
@@ -0,0 +1,404 @@
+    // TODO: add a proof of possession/knowledge?
+    pub vss_pk: PublicPoly<G>,
+}
+
+/// A complaint/fraud claim against a dealer that created invalid encrypted share.
+// TODO: add Serialize & Deserialize.
+#[derive(Clone, PartialEq, Eq)]
+pub enum Complaint<EG: GroupElement> {
+    /// The identity of the sender.
+    NoShare(ShareIndex),
+    /// The identity of the sender and the recovery package.
+    // An alternative to using ECIES & ZKPoK for complaints is to use different ECIES public key
+    // for each sender, and in case of a complaint, simply reveal the relevant secret key.
+    // This saves the ZKPoK with the price of publishing one ECIES public key & PoP for each party,
+    // resulting in larger communication in the happy-path.
+    InvalidEncryptedShare(ShareIndex, RecoveryPackage<EG>),
+}
+
+/// A [DkgSecondMessage] is sent during the second phase of the protocol. It includes complaints
+/// created by receiver of invalid encrypted shares.
+#[derive(Clone, PartialEq, Eq)]
+pub struct SecondMessage<EG: GroupElement> {
+    pub sender: ShareIndex,
+    /// List of complaints against other parties. Empty if there are none.
+    pub complaints: Vec<Complaint<EG>>,
+}
+
+/// Mapping from node id to the share received from that leader.
+pub type SharesMap<G> = HashMap<ShareIndex, <G as GroupElement>::ScalarType>;
+
+/// [DkgOutput] is the final output of the DKG protocol in case it runs
+/// successfully. It can be used later with [ThresholdBls], see examples in tests.
+#[derive(Clone, Debug)]
+pub struct DkgOutput<G: GroupElement, EG: GroupElement> {
+    pub nodes: Nodes<EG>,
+    pub vss_pk: Poly<G>,
+    pub share: Share<G::ScalarType>,
+}
+
+/// A dealer in the DKG ceremony.
+///
+/// Can be instantiated with G1Curve or G2Curve.
+impl<G: GroupElement, EG: GroupElement> Party<G, EG>
+where
+    <G as GroupElement>::ScalarType: Serialize + DeserializeOwned,
+    EG: Serialize,
+    <EG as GroupElement>::ScalarType: HashToGroupElement,
+{
+    /// 1. Create a new ECIES private key and send the public key to all parties.
+    /// 2. After all parties have sent their ECIES public keys, create the set of nodes.
+    /// 3. Create a new Party instance with the ECIES private key and the set of nodes.
+    pub fn new<R: AllowedRng>(
+        ecies_sk: ecies::PrivateKey<EG>,
+        nodes: Nodes<EG>,
+        threshold: u32, // The number of parties that are needed to reconstruct the full signature.
+        random_oracle: RandomOracle,
+        rng: &mut R,
+    ) -> Result<Self, FastCryptoError> {
+        let ecies_pk = ecies::PublicKey::<EG>::from_private_key(&ecies_sk);
+
+        // Check if the public key is in one of the nodes.
+        let curr_node = nodes
+            .iter()
+            .find(|n| n.pk == ecies_pk)
+            .ok_or(FastCryptoError::InvalidInput)?;
+
+        // Generate a secret polynomial and commit to it.
+        let vss_sk = PrivatePoly::<G>::rand(threshold - 1, rng);
```

Comment:
do we need to check that threshold is in the expected range ie <= n?

Reply:
added a check